### ISSUE TITLE
Using Non-Managed Apache Iceberg Tables in Snowflake.

### ISSUE THREAD

**Issue by j3-signalroom** [j3-signalroom/apache_flink-kickstarter#521]

**Comment by j3-signalroom** [comment]
Research links:

- https://medium.com/snowflake/an-overview-of-snowflake-apache-iceberg-tables-d5e85864ac99
- https://quickstarts.snowflake.com/guide/getting_started_iceberg_tables/index.html#0
- https://docs.snowflake.com/en/user-guide/tables-iceberg
- https://www.snowflake.com/en/blog/introducing-polaris-catalog/
-

**Comment by j3-signalroom** [comment]
Creating a non-managed Apache Iceberg table for Snowflake involves defining the table externally (outside Snowflake) and using Snowflake to query the table's data. A non-managed table means the data remains in its original external location (e.g., Amazon S3, Azure Blob Storage, Google Cloud Storage) rather than being managed by Snowflake.

## **1. Add Metadata Integration with Snowflake**
In Snowflake, define an **external table** pointing to the Iceberg table's location:

### a. **Create an External Stage**
```sql
CREATE STAGE my_stage
  URL = 's3://your-bucket/warehouse/db_name/table_name/'
  STORAGE_INTEGRATION = my_storage_integration;
```

### b. **Create an External Table in Snowflake**
Create the external table with a schema that matches the Iceberg table:

```sql
CREATE EXTERNAL TABLE my_iceberg_table (
  id BIGINT,
  data STRING
)
LOCATION = '@my_stage';
```

This assumes you’ve configured the `STORAGE_INTEGRATION` in Snowflake to handle access to the external storage.

## **2. Query the Table**
After creating the external table in Snowflake, you can query it as a standard Snowflake table:

```sql
SELECT * FROM my_iceberg_table;
```

---

## **3. Maintain the Table**
- Use Apache Iceberg's utilities for schema evolution, partitioning, and compaction.
- Ensure external metadata is consistent with Snowflake’s external table definition.

---

This setup allows Snowflake to act as a querying layer for non-managed Iceberg tables while the data resides externally.

**Comment by j3-signalroom** [comment]
In Snowflake, the `STORAGE_INTEGRATION` parameter specifies a named integration object that allows Snowflake to interact securely with external cloud storage systems (e.g., AWS S3). The `STORAGE_INTEGRATION` must be pre-configured in Snowflake and is associated with a specific cloud storage provider.

## **Amazon S3**
- **Configuration:**
  - `STORAGE_INTEGRATION` is associated with an IAM Role in AWS.
  - The IAM Role must have appropriate permissions (e.g., `s3:GetObject`).
- **Example:**
  ```sql
  CREATE STORAGE INTEGRATION my_s3_integration
    TYPE = EXTERNAL_STAGE
    STORAGE_PROVIDER = 'S3'
    ENABLED = TRUE
    STORAGE_AWS_ROLE_ARN = 'arn:aws:iam::123456789012:role/my-s3-role'
    STORAGE_ALLOWED_LOCATIONS = ('s3://my-bucket/my-path/');
  ```

---

## **Retrieving Valid Storage Integrations**
You can view existing `STORAGE_INTEGRATION` objects in your Snowflake account using:

```sql
SHOW INTEGRATIONS;
```

**Comment by j3-signalroom** [comment]
Terraform configuration:
- https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/storage_integration
- https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/stage
- https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/external_table

**Comment by j3-signalroom** [comment]
I reopened the issue because Snowflake is not recognizing the Apache Iceberg Tables.

**Comment by j3-signalroom** [comment]
<img width="1942" alt="image" src="https://github.com/user-attachments/assets/98926a02-840b-4e84-9474-e77948fb10c9">

**Comment by j3-signalroom** [comment]
<img width="1933" alt="image" src="https://github.com/user-attachments/assets/c4770739-c82d-4051-822c-5d3c7fa8e055">

**Comment by j3-signalroom** [comment]
<img width="1678" alt="image" src="https://github.com/user-attachments/assets/d98eeeb4-932c-467a-8090-0f8a5e13a8e5" />

**Comment by j3-signalroom** [comment]
![image](https://github.com/user-attachments/assets/40c928f6-7671-45e7-a623-d61807642c28)

**Comment by j3-signalroom** [comment]
![image](https://github.com/user-attachments/assets/9dfd64a0-a1cb-4974-acd1-b6acf552822b)

**Comment by j3-signalroom** [comment]
Issue resolved in #549 .
